proposals/Level-1+N.md

```diff
@@ -0,0 +1,204 @@
+end
+```
+
+One downside of this form is that it does not directly support running different
+code for different exception types. This can still be accomplished, however, by
+dyanmically inspecting the exception inside the catch block.
+
+# Level 3
+
+## Resumable Exceptions
+
+This proposal allows exception handlers to return a value to the site where the
+exception was thrown. Resumable exceptions enable or are equivalent to a number
+of advanced control flow features such as generators, coroutines, delimited
+continuations and effect handlers.
+
+In high level code, this would enable programs such as:
+
+```
+Exception FileNotFound: string -> string
+
+try {
+  filename = "foo.txt"
+  while(!(file = open(filename))) {
+    filename = throw FileNotFound(filename);
+  }
+} catch (FileNotFound e) {
+  new_file_name = ask_user_for_filename();
+  resume e(new_file_name)
+}
+```
+
+The key changes are that exception types are now allowed to include return
+values, `throw` can return a value (we might want to introduce a different
+opcode for this version, such as `raise` or `throw_resumable`), and there is a
+`resume` instruction that returns values back to where the exception was thrown
+from.
+
+Below are several examples showing how other control flow patterns can be
+expressed in terms of resumable exceptions.
+
+### Example: Algebraic Effect Handlers
+
+Consider the following example, taken from http://www.eff-lang.org/try/:
+
+```
+handle
+    std#print "A";
+    std#print "B";
+    std#print "C";
+    std#print "D"
+with
+| std#print msg k ->
+    std#print ("I see you tried to print " ^ msg ^ ". Okay, you may.\n");
+    k ()
+```
+
+This could be lowered to Wasm resumable exception pseudo-code like this:
+
+```
+Exception StdPrint: string -> ()
+
+import function RealPrint: string -> ()
+
+try:
+  throw StdPrint("A")
+  throw StdPrint("B")
+  throw StdPrint("C")
+  throw StdPrint("D")
+catch e: StdPrint(msg):
+  RealPrint("I see you tried to print " + msg + ". Okay, you may.\n")
+  resume e()
+end
+```
+
+### Example: Coroutines
+
+Consider the example below taken from https://en.wikipedia.org/wiki/Coroutine:
+
+```
+var q := new queue
+
+coroutine produce
+    loop
+        while q is not full
+            create some new items
+            add the items to q
+        yield to consume
+
+coroutine consume
+    loop
+        while q is not empty
+            remove some items from q
+            use the items
+        yield to produce
+```
+
+Using similar high level pseudo code, we can convert this as follows to use
+resumable exceptions:
+
+```
+Exception yield: () -> ()
+
+function produce():
+  loop:
+    while q is not full:
+      create some new items
+      add the items to q
+    throw yield()
+
+function consume():
+  loop:
+    while q is not empty:
+      remove some items from q
+      use the items
+    throw yield()
+
+function run():
+  // Start the producer
+  c1 = try:
+    produce()
+  catch e: yield():
+    e
```

Comment:
Nit: May be clearer if the assignment to c1 occurs inside here (same below).

Reply:
Good idea. Done.